Map Page using REST APIs

deybyr647 opened the pull request:
### Changes to be merged into dev branch
_Refer to [pages/map/index.tsx](https://github.com/deybyr647/RecycleIT/blob/map-with-rest-apis/pages/map/index.tsx)_
- Map page now shows a map by default, instead of only when there are coordinates present

- Zip code input/search now fully functional

- Implemented the Google Maps Place Search API, which helps in searching for nearby recycling centers using coordinates provided from user geolocation or zip code

- Implemented Google Maps Geocoding REST API, for use with zip code searches. The API transforms the zip code to coordinates used by the Google Maps Place Search REST API.

- Custom markers now render properly to the map, using place coordinates provided by Place Search API

- Markers using user coordinates are red, Markers using zip code coordinates are blue, and recycling center markers are green

### Upcoming additions/features & notes

- Note: I added another column in the map page, which will show some basic information for each of the places shown on the map. For now, it's just a placeholder visual component.

- Future feature: Interactive markers or interactive place info cards? Info cards could show a button which would emphasize the marker for the selected place on the map upon being clicked & vice versa (if possible).

- Dynamic routes? Place info cards could show a hyper link or button which would take the user to a dynamic route, where the user would be able to see more thorough place information. This would be because the Place Search API only provides some basic stuff. Place ID could be passed as a URL query parameter, or thru local storage or some other way, and would be used in the dynamic page/route to make calls to the Google Place Details REST API. This dynamic route/page would then show place information based on JSON responses, like the main Map page.

-  Way to display zip code area boundaries?

deybyr647 committed:
Began implementing place markers. Added a JSON file with an example response for testing, and added axios as a dep.

deybyr647 committed:
Added a proxy for use with REST APIs, and created a custom server.js for serving over https in localhost

deybyr647 committed:
Modified page to show map, with center of the map being dependent on presence of user coordinates (shows a map regardless). Moved API logic out of Map component, for future use with information components. APIs will be called based on changes in coordinates (useEffect hook implementation)

deybyr647 committed:
Began implementing zip code usage. Renamed some functions for better readability & understanding of code

deybyr647 committed:
Minor edit

deybyr647 committed:
Implemented a new layout on map page & Finalized
API functionality (for now).

deybyr647 commented:
Hey @stephenbunch ! How are you? I hope you are well. Here's my next pull request, which will be merged to the dev branch. Feel free to review my code & let me know how it's looking when you get the chance :)

-- Deyby

stephenbunch commented on pages/map/index.tsx:
```diff
@@ -48,52 +51,106 @@ const Marker = ({color}: MarkerProps) => (
+        setZip("");
+    }
+
+    const getPlaceCoordsWithUserCoords = async () => {
```
one thing you can do to improve readability is to refactor some of these functions into [pure functions](https://en.wikipedia.org/wiki/Pure_function). for example, the async getters can be moved outside the component altogether which would condense the component code and make it easier to read. otherwise, it's hard to tell which variables are used by which functions and what state is modified as a result of executing those functions. it also makes the functions themselves harder to reuse because they're mixed with side effects.

- `getPlaceCoordsWithUserCoords` could be refactored into a helper method that takes `userCoords` and returns a list of places.
- `getPlaceCoordsWithZip` could be refactored into a helper that takes `zip` and returns the geocoded result.

then, the `useEffect` block below could be rewritten as:

```
useEffect(async () => {
  setPlaces(await getPlaceCoordsWithUserCoords(userCoords));
}, [userCoords]);
```

and the `submitHandler` above could be rewritten as:

```
const submitHandler = async (e: any) => {
  e.preventDefault();
  
  const geocodedResult = await getPlaceCoordsWithZip(zip);
  setZipStatus(true);

  setUserCoords(/* same code as before */);
}
```

with this, now you don't need to duplicate the place fetching logic since `setUserCoords` will cause the `useEffect` block to rerun.

deybyr647 replied:
I gave the refactoring thing a try, where I called functions within the state setters. Doing that gave me runtime errors, where Next said something like ```(Object...) is undefined```
async callbacks & promises aren't a thing in useEffect because of race conditions apparently, but async functions can be called from within useEffect, which is pretty much what I'm already doing. Here's what I mean:

Bad:
`
useEffect(async () => {
})`

Good:

`const x = async () => {
        let z = await y
}`
`useEffect(() => {
        x()
})`

The same applies to using async event handlers like you suggested, although event.persist could probably be used. Here's some stuff I found about this:
[async useEffect](https://stackoverflow.com/questions/53332321/react-hook-warnings-for-async-function-in-useeffect-useeffect-function-must-ret)
[async useEffect GitHub issue](https://github.com/facebook/react/issues/14326)
[async event handlers](https://medium.com/@ian.mundy/async-event-handlers-in-react-a1590ed24399)

I could probably store the return values of the async getters in variables & then use the variables or properties of those variables to set states:

`const getPlacesWithUserCoords = async () => {foo: "bar", a: "b", c: [1, 2, 3, 4, 5]}`
`const placesReq = getPlacesWithUserCoords()`
`setPlaces(placesReq.c)`

stephenbunch replied:
ah, that's too bad. I hate it when framework authors skimp on async design. what they should have done is created an API that makes it possible to cancel the previous effect if the effect is rerun before the async task completes. the reality is, you still have to deal with this problem regardless of whether the async task is passed directly to `useEffect` or run inside a sync callback.

regarding the runtime errors from Next, you should debug it some more to figure out why the error is occurring. I recommend [chrome dev tools](https://developers.google.com/web/tools/chrome-devtools/javascript) for debugging javascript. I also found an article specifically on [debugging Next](https://nextjs.org/docs/advanced-features/debugging). if you can't figure it out after an hour or so, we can go over it on Saturday. debugging is definitely a skill you'll want to get good at. I've spent just as many hours debugging as I have writing code. no one enjoys it, but it's a brute-force way of figuring out a mistake.

deybyr647 replied:
Hey @stephenbunch . I was able to refactor the component. I moved all the API logic to a separate file altogether.  Everything is still working as intended, but there is now a delay in the marker loading, due to the async tasks. What I can do to combat this, as to not have a user leave the site, is display a loading spinner, and control it with a loading state. Either way, check this [commit](https://github.com/deybyr647/RecycleIT/pull/3/commits/fba720e253cb19f84dd6396a820eceee198fb4e0) whenever you can, and let me know what you think!

stephenbunch commented on server/server.js:
```diff
@@ -0,0 +1,23 @@
+    if (err) throw err;
+    console.log("> Server started on https://localhost:3000");
+  });
+});
```
it's generally good practice to insert a newline at the end of each file.

https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

[editorconfig](https://editorconfig.org/) is a great tool that can auto-insert newlines on save. it can also configure other annoyances like tabbing defaults for different file types.

deybyr647 replied:
I think I was able to correct this by editing the .gitconfig file in my system. Did so by changing core.autocrlf from input to true (on Widnows).

deybyr647 committed:
Changed CRLF setting globally

deybyr647 committed:
Refactored MapPageContent component by moving
async API logic into a new file (util.ts). Also added
"open" package to automate web page opening, similar to "yarn start"
in create-reat-app. Removed data JSON file, since API now works fine
